File: libmuscle/python/libmuscle/snapshot.py
from abc import ABC, abstractmethod
from dataclasses import dataclass
from typing import Dict, List, Optional, cast

import msgpack
from ymmsl import Reference, Settings

from libmuscle.mpp_message import MPPMessage
from libmuscle import communicator


class Snapshot(ABC):
    """Snapshot data structure.

    This is an abstract base class, implementations are provided by subclasses.
    """
    SNAPSHOT_VERSION_BYTE = b'\0'

    def __init__(self,
                 triggers: List[str],
                 wallclock_time: float,
                 port_message_counts: Dict[str, List[int]],
                 is_final_snapshot: bool,
                 message: Optional['communicator.Message'],
                 settings_overlay: Settings) -> None:
        self.triggers = triggers
        self.wallclock_time = wallclock_time
        self.port_message_counts = port_message_counts
        self.is_final_snapshot = is_final_snapshot
        self.message = message
        # self.message is None for implicit snapshots, so we cannot store the
        # Settings overlay in that message object.
        self.settings_overlay = settings_overlay

    @classmethod
    @abstractmethod
    def from_bytes(cls, data: bytes) -> 'Snapshot':
        """Create a snapshot object from binary data.

        Args:
            data: binary data representing the snapshot. Note that this must
                **exclude** the versioning byte.
        """
        ...

    @abstractmethod
    def to_bytes(self) -> bytes:
        """Convert the snapshot object to binary data.

        Returns:
            Binary data representing the snapshot. Note that this must
                **exclude** the versioning byte.
        """
        ...


class MsgPackSnapshot(Snapshot):
    """Snapshot stored in messagepack format
    """
    SNAPSHOT_VERSION_BYTE = b'1'

    @classmethod
    def from_bytes(cls, data: bytes) -> 'Snapshot':
        dct = msgpack.loads(data)
        return cls(dct['triggers'],
                   dct['wallclock_time'],
                   dct['port_message_counts'],
                   dct['is_final_snapshot'],
                   cls.bytes_to_message(dct['message']),
                   Settings(dct['settings_overlay']))

    def to_bytes(self) -> bytes:
        return cast(bytes, msgpack.dumps({
            'triggers': self.triggers,
            'wallclock_time': self.wallclock_time,
            'port_message_counts': self.port_message_counts,
            'is_final_snapshot': self.is_final_snapshot,
            'message': self.message_to_bytes(self.message),
            'settings_overlay': self.settings_overlay.as_ordered_dict()
        }))

    @staticmethod
    def message_to_bytes(message: Optional['communicator.Message']) -> bytes:
        """Use MPPMessage serializer for serializing the message object
        """
        if message is None:
            return b''
        settings = Settings()
        if message.settings is not None:
            settings = message.settings
        return MPPMessage(Reference('_'), Reference('_'), None,
                          message.timestamp, message.next_timestamp,
                          settings, 0, -1.0, message.data).encoded()

    @staticmethod
    def bytes_to_message(data: bytes) -> Optional['communicator.Message']:
        """Use MPPMessage deserializer for serializing the message object
        """
        if not data:
            return None
        mpp_message = MPPMessage.from_bytes(data)
        return communicator.Message(mpp_message.timestamp,
                                    mpp_message.next_timestamp,
                                    mpp_message.data,
                                    mpp_message.settings_overlay)


@dataclass
class SnapshotMetadata:
    """Metadata of a snapshot for sending to the muscle_manager.
    """
    triggers: List[str]
    wallclock_time: float
    timestamp: float
    next_timestamp: Optional[float]
    port_message_counts: Dict[str, List[int]]
    is_final_snapshot: bool
    # storing as str, because Path cannot be serialized by msgpack
    snapshot_filename: str

    @staticmethod
    def from_snapshot(snapshot: Snapshot, snapshot_filename: str
                      ) -> 'SnapshotMetadata':
        """Create snapshot metadata from the given snapshot and filename
        """
        return SnapshotMetadata(
            snapshot.triggers,
            snapshot.wallclock_time,
            snapshot.message.timestamp if snapshot.message else float('NaN'),
            snapshot.message.next_timestamp if snapshot.message else None,
            snapshot.port_message_counts,
            snapshot.is_final_snapshot,
            snapshot_filename
        )

File: libmuscle/cpp/src/libmuscle/snapshot.cpp
#include <libmuscle/snapshot.hpp>

#include <cmath>

#include <msgpack.hpp>

#include <libmuscle/mcp/data_pack.hpp>
#include <libmuscle/mpp_message.hpp>


namespace libmuscle { namespace impl {

Snapshot::Snapshot(
            std::vector<std::string> const & triggers,
            double wallclock_time,
            std::unordered_map<std::string, std::vector<int>> const & port_message_counts,
            bool is_final_snapshot,
            Optional<Message> const & message,
            ::ymmsl::Settings const & settings_overlay
            )
        : triggers(triggers)
        , wallclock_time(wallclock_time)
        , port_message_counts(port_message_counts)
        , is_final_snapshot(is_final_snapshot)
        , message(message)
        , settings_overlay(settings_overlay)
    {}

Snapshot Snapshot::from_bytes(DataConstRef const & data) {
    // decode
    auto zone = std::make_shared<msgpack::zone>();
    DataConstRef dict = mcp::unpack_data(zone, data.as_byte_array(), data.size());

    // convert lists/dicts to vectors/unordered_maps
    std::vector<std::string> triggers;
    auto data_triggers = dict["triggers"];
    for (std::size_t i=0; i<data_triggers.size(); ++i) {
        triggers.push_back(data_triggers[i].as<std::string>());
    }

    std::unordered_map<std::string, std::vector<int>> port_message_counts;
    auto data_pmc = dict["port_message_counts"];
    for (std::size_t i=0; i<data_pmc.size(); ++i) {
        std::vector<int> counts;
        for (std::size_t j=0; j<data_pmc.value(i).size(); ++j) {
            counts.push_back(data_pmc.value(i)[j].as<int>());
        }
        port_message_counts[data_pmc.key(i)] = counts;
    }

    Optional<Message> message;
    if (!dict["message"].is_nil()) {
        auto mpp_message = MPPMessage::from_bytes(dict["message"]);
        message = Message(mpp_message.timestamp, mpp_message.data);
        if (mpp_message.next_timestamp.is_set()) {
            message.get().set_next_timestamp(mpp_message.next_timestamp.get());
        }
        if (!mpp_message.settings_overlay.is_nil()) {
            message.get().set_settings(
                    mpp_message.settings_overlay.as<::ymmsl::Settings>());
        }
    }

    return Snapshot(
            triggers,
            dict["wallclock_time"].as<double>(),
            port_message_counts,
            dict["is_final_snapshot"].as<bool>(),
            message,
            dict["settings_overlay"].as<::ymmsl::Settings>()
            );
}

DataConstRef Snapshot::to_bytes() const {
    Data d_triggers = Data::nils(triggers.size());
    for (std::size_t i=0; i<triggers.size(); ++i) {
        d_triggers[i] = triggers[i];
    }

    Data pmc = Data::dict();
    for (const auto & kv : port_message_counts) {
        Data counts = Data::nils(kv.second.size());
        for (std::size_t i=0; i<kv.second.size(); ++i) {
            counts[i] = kv.second[i];
        }
        pmc[kv.first] = counts;
    }

    Data dict;
    // Note setting dict in two branches, to avoid a memcopy of the encoded MMPMessage
    if (message.is_set()) {
        auto msg = message.get();
        MPPMessage mpp_msg(
                "_",
                "_",
                {},
                msg.timestamp(),
                msg.has_next_timestamp() ? msg.next_timestamp() : Optional<double>(),
                msg.has_settings() ? msg.settings() : Data(),
                0,
                -1.0,
                msg.data());
        // Initializing a Data::dict with a DataConstRef is allowed, but assignment
        // after creation is not
        dict = Data::dict(
            "triggers", d_triggers,
            "wallclock_time", wallclock_time,
            "port_message_counts", pmc,
            "is_final_snapshot", is_final_snapshot,
            "message", mpp_msg.encoded(),
            "settings_overlay", Data(settings_overlay));
    } else {
        dict = Data::dict(
            "triggers", d_triggers,
            "wallclock_time", wallclock_time,
            "port_message_counts", pmc,
            "is_final_snapshot", is_final_snapshot,
            "message", Data(),
            "settings_overlay", Data(settings_overlay));
    }

    msgpack::sbuffer sbuf;
    msgpack::pack(sbuf, dict);

    auto bytes = Data::byte_array(sbuf.size());
    memcpy(bytes.as_byte_array(), sbuf.data(), sbuf.size());

    return bytes;
}

SnapshotMetadata::SnapshotMetadata(
            std::vector<std::string> const & triggers,
            double wallclock_time,
            double timestamp,
            Optional<double> next_timestamp,
            std::unordered_map<std::string, std::vector<int>> const & port_message_counts,
            bool is_final_snapshot,
            std::string const & snapshot_filename)
        : triggers(triggers)
        , wallclock_time(wallclock_time)
        , timestamp(timestamp)
        , next_timestamp(next_timestamp)
        , port_message_counts(port_message_counts)
        , is_final_snapshot(is_final_snapshot)
        , snapshot_filename(snapshot_filename)
    {}

SnapshotMetadata SnapshotMetadata::from_snapshot(
        Snapshot const & snapshot, std::string const & snapshot_filename) {
    double timestamp = std::numeric_limits<double>::quiet_NaN();
    Optional<double> next_timestamp;
    if (snapshot.message.is_set()) {
        timestamp = snapshot.message.get().timestamp();
        if (snapshot.message.get().has_next_timestamp()) {
            next_timestamp = snapshot.message.get().next_timestamp();
        }
    }
    return SnapshotMetadata(
            snapshot.triggers,
            snapshot.wallclock_time,
            timestamp,
            next_timestamp,
            snapshot.port_message_counts,
            snapshot.is_final_snapshot,
            snapshot_filename);
}

} }
